File: NeuralNetwork/HiddenNode.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography.X509Certificates;
using System.Threading.Tasks;
using GradientDescentApp;

namespace GradientDescentApp
{
    public class HiddenNode : Node
    {
        public float bias;
        protected List<Weight> outgoingWeights;
        protected List<Weight> incomingWeights;
        

        public HiddenNode(float bias)
        {
            this.bias = bias;
            outgoingWeights = new List<Weight>();
            incomingWeights = new List<Weight>();
        }

        public void AddOutgoingWeight(Weight weight)
        {
            outgoingWeights.Add(weight);
        }
        public void AddIncomingWeight(Weight weight)
        {
            incomingWeights.Add(weight);
        }

        //z1 = w1x1 + w2x2 + b
        public void Activation()
        {
            float activation = bias;

            foreach(Weight weight in incomingWeights)
            {
                activation += weight.value * weight.prev.value;
            }

            activation = MathF.Max(0, activation); // ReLU activation function

            this.value = activation;
        }

        public float HiddenNodeError()
        {
            float error = 0;
            foreach(Weight weight in outgoingWeights)
            {
                error += weight.value * weight.next.error; 
            }
            this.error = error;
            return error;
        }
    }
}

File: NeuralNetwork/Python/HiddenNode.py
from Node import Node
"""
The Hidden Node. Used in the hidden layers of the neural network
between the input and output layers. Each node has an activation function which is 
w1x1 + w2x2 + ... + b, where w is the weight of the connection, 
x is the x value, and b is the bias. 
The equation is then passed through a ReLU function to see if the neuron "fires" or not.
"""

class HiddenNode(Node):

    def __init__(self, bias):
        super().__init__()
        self.bias = bias

    # z1 = w1x1 + w2x2 + b
    def Activation(self):
        activation = self.bias

        for weight in self.incomingWeights:
            activation += weight.value * weight.prev.value

        activation = max(0, activation)  # ReLU activation function

        self.value = activation

    def HiddenNodeError(self):
        error = 0.0
        for weight in self.outgoingWeights:
            error += weight.value * weight.next.error

        self.error = error
        return error
